File: src/main/java/sx/Dispose.java
package sx;

import com.google.common.base.Stopwatch;
import org.apache.commons.logging.Log;
import org.apache.commons.logging.LogFactory;

/**
 * Created by masc on 07.07.15.
 */
public class Dispose {
    static Log mLog = LogFactory.getLog(Dispose.class);

    /**
     * Customized disposal
     * @param r Dispose operation
     */
    public static void safely(final Runnable r, final Action<Exception> exceptionHandler) {
        try {
            r.run();
        } catch (Exception e) {
            if (exceptionHandler != null)
                exceptionHandler.perform(e);
        }
    }

    /**
     * Safely disposes an instance implementing @link Disposable, logging, timing and errors
     * @param d Disposable instance
     * @param exceptionHandler Optional exception handler
     */
    public static void safely(final Disposable d, final Action<Exception> exceptionHandler) {
        mLog.info(String.format("Disposing [%s]", d.getClass().getName()));
        Stopwatch sw = Stopwatch.createStarted();
        safely(new Runnable() {
            @Override
            public void run() {
                d.close();
            }
        }, new Action<Exception>() {
            @Override
            public void perform(Exception e) {
                mLog.error(e.getMessage(), e);
                exceptionHandler.perform(e);
            }
        });
        mLog.info(String.format("Disposed [%s] in [%s]", d.getClass().getName(), sw.toString()));
    }

    /**
     * Safely disposes an instance implementing @link Disposable, logging timing and errors
     * @param d
     */
    public static void safely(final Disposable d) {
        safely(d, null);
    }
}

File: src/main/java/sx/Dispose.kt
package sx

import org.slf4j.LoggerFactory

/**
 * Created by masc on 07.07.15.
 */
object Dispose {
    internal var log = LoggerFactory.getLogger(Dispose::class.java)

    /**
     * Customized disposal
     * @param r Dispose operation
     */
    @JvmOverloads fun safely(r: () -> Unit, exceptionHandler: ((Exception) -> Unit)? = null) {
        try {
            r()
        } catch (e: Exception) {
            exceptionHandler?.invoke(e)
        }
    }

    /**
     * Safely disposes an instance implementing @link Disposable, logging, timing and errors
     * @param d Disposable instance
     * @param exceptionHandler Optional exception handler
     */
    @JvmOverloads fun safely(d: Disposable, exceptionHandler: ((Exception) -> Unit)? = null) {
        log.info(String.format("Disposing [%s]", d.javaClass.name))
        val sw = Stopwatch.createStarted()
        this.safely(
                { d.close() },
                { e ->
                    log.error(e.message, e)
                    exceptionHandler?.invoke(e)
                })
        log.info(String.format("Disposed [%s] in [%s]", d.javaClass.name, sw.toString()))
    }

    /**
     * Safely disposes an instance implementing @link Disposable, logging, timing and errors
     * @param d Disposable instance
     * @param exceptionHandler Optional exception handler
     */
    @JvmOverloads fun safely(d: sx.legacy.Disposable, exceptionHandler: ((Exception) -> Unit)? = null) {
        this.safely(object : Disposable {
            override fun close() {
                d.close()
            }
        }, exceptionHandler)
    }
}
